### DISCUSSION TITLE
Which objects are proper to compute the equations (Auxvariable, postprocessor, etc)?

### DISCUSSION THREAD

**Discussion by hokonkim** [idaholab/moose#19381] [Q&A General] [answered]
Hello, I would like to compute the equations like below.

![image](https://user-images.githubusercontent.com/46009457/141668135-1bf27473-4e7a-45ea-b347-cb50b4e012f0.png)

From several dicussions I made at here, I knew

V_Em: local volume of Element,
          can be computed by VolumeAux (recently made)

But, I'm confused about computing dV_cr/dt *delta_t and V_act.

dV_cr/dt has parameters such as S_D,Cr2O3 (changeable) * V_Em (constant) * V_mol_Cr2O3 (constant) * 1/zF (constant).

In S_D,Cr2O3, V_act (changeable), V_Em(constant), and eta (changeable, AuxVariable).

**To compute V_cr/dt*delta_t and V_act, which objects are needed?**

I tried to compute as Material Properties.. because it can use Auxvariable and Postprocessors, but I'm stuck..
(struggling to implement V_cr_last, delta_t, and V_act)

How can I do compute this?

    [Materials]
      [./CrDeposition]
        type  = ParsedMaterial
        block = 'PT_TPB_TET4'
        f_name = 'CrDepoActivity_SD'
        constant_names        = 'R          T        z     F          x_CrO2OH2   x_H2O  V_molar_Cr2O3'
                            # (J/K/mol),  (K), (# of e-), (C/mol),    (non),      (non), (cm^3)
        constant_expressions  = '8.3144598  1073.0   4   96485.3329   1e-8        0.01   29.12'
        function = 'V_cr_last +
                    (274.626 * VolumeCoverage * x_CrO2OH2^(0.5) * x_H2O^(0.5) * 2 * sinh(F * 2 / R / T * aux_eta_tpb)) *
                    aux_vol_em * V_molar_Cr2O3 / z / F'
        args = 'aux_eta_tpb aux_vol_em'
        postprocessor_names = 'V_cr_last VolumeCoverage'
        # Make values see at output(*.e) file
        # outputs = exodus
      [../]
    []

    [AuxVariables]
      [./aux_vol_em]
        order = CONSTANT
        family = MONOMIAL
        block = 'PT_TPB_TET4'
      [../]
    []

    [AuxKernels]
      [./volume_element]
        type = VolumeAux
        variable = aux_vol_em
        block = 'PT_TPB_TET4'
      [../]
    []

Also, 

![image](https://user-images.githubusercontent.com/46009457/141669875-f1e389cb-375a-46ab-9fb2-0420af6160a5.png)

**I'm not sure I need "Material properties" in AuxKernel expression like this.
(Material properties are currently not supported, but it would be really easy to add it so feel free to contact us.)**

But, how can I do this?

**Lastly, what is the best way to compute delta t?**

Thanks.

Sincerely,
Hokon Kim

**Comment by GiudGiud** [comment, marked as answer]
> To compute V_cr/dt*delta_t and V_act, which objects are needed?

Is V_cr a main variable? If so you just need to form equation number 3 with kernels. Equation number 2 will naturally happen with an Euler time stepper (default is Implicit Euler)

V_act seems fairly simple to compute from 2 variables, it s just a difference. A ParsedAux will do it. 

> I'm not sure I need "Material properties" in AuxKernel expression like this.

You pretty much have a choice between using AuxVariables and MaterialProperties, unless you are using Automatic Differentiation in which case you should prefer to use Material Properties. 
Use [Materials] to compute Material properties, and [AuxKernels] to compute aux-variables. There's parsed formulations for both. 

> Lastly, what is the best way to compute delta t?

delta t is usually accessible using simply `_dt`. 
If not, it can be retrieved from the problem object with `_problem_ptr->dt()`

**Reply by hokonkim** [reply]
So far, I dealt V_cr as a AuxVariable. I'm confused about which objects are proper for my application.

Kernel is needed to do something related to physics for reaction, as far as I know.

Do you think dealing equation 3 (dV_cr/dt = S_D * V_Em * ...) as Kernel would be proper for my purpose?

**Reply by GiudGiud** [reply]
A kernel is usually used for each term of the equation. So dV_cr/dt is a TimeDerivative kernel on variable V_cr, and the other term is another kernel (you ll have to add that one, there is no parsedKernel afaik, only a parsedAux)

Note that kernels are used to describe equations you want to solve using a numerical solver. If the equations are made up of terms you already know from a different solve, then you use AuxKernels to describe them, since it s not a real solve, just a substitution

If V_cr is not a nonlinear variable (eg a main variable, that you want to solve for), then my answer here will help you create the auxkernel you need https://github.com/idaholab/moose/discussions/19292

**Reply by hokonkim** [reply]
In this case, I don't know V_Cr unless I compute the V_cr.

So, V_Cr needs to be dealt as Variable. 

At a previous discussion, local volume of Element can be access with

     _current_elem_volume

(how did you make a block for only this word?)

In Equation.3, S_D,Cr2O3 needs to be updated with (V_Em - V_Cr) / V_Em.

Even though I can use V_em as _current_elem_volume in ComputeQpresidual, 

how can I V_cr in ComputeQpresidual?

Still.. I have a lot on my mind.

**Reply by GiudGiud** [reply]
If the kernel is for variable V_cr then you can access its value using _u[_qp] . _u is the predefined name for the variable that the kernel applies to. It s defined for most kernels. 

If the kernel is for another variable, you will need to add V_cr as a coupled variable in the kernel before being able to access it. See CoupledForce kernel for example
